hanabi_lib/canonical_encoders.h

```diff
@@ -29,8 +29,8 @@ namespace hanabi_learning_env {
 // This is the canonical observation encoding.
 class CanonicalObservationEncoder : public ObservationEncoder {
  public:
-  explicit CanonicalObservationEncoder(const HanabiGame* parent_game)
-      : parent_game_(parent_game) {}
+  explicit CanonicalObservationEncoder(const HanabiGame* parent_game, bool using_joint_obs)
```

Comment:
same as before, we can put the logic in the environment by calling difference encode function, instead of adding additional parameter in the constructor.

Reply:
The main issue I had with doing it the way you propose was the use of `CanonicalObservationEncoder::Shape`, which is called by `featureSize` in `hanabi_env.h`, and whose result depends on joint vs. not-joint type. Some workarounds could be: 

- Modify `Shape` function, but this is even more problematic than what I did before, since it's an override function. 
- Add a new function `CanonicalObservationEncoder::ShapeJointObs` and use this one. Maybe this is the best option?
- Compute the new shape in `hanabi_env.h`, but I think this is ugly. 

Any preference or other suggestions?

Reply:
“Add a new function CanonicalObservationEncoder::ShapeJointObs and use this one. Maybe this is the best option?”

I would prefer this one! Thanks